pkg/roll/run_update.go

```diff
@@ -238,10 +230,16 @@ func (u update) lockRCs() error {
 	}
 
 	err = u.lock.Lock(oldPath)
+	if err != nil {
+		// The second key couldn't be locked, so release the first key before retrying.
+		RetryOrQuit(
+			func() error { return u.lock.Unlock(newPath) },
+			done,
+			u.logger,
+			fmt.Sprintf("unlocking %s", newPath),
+		)
+	}
 	if _, ok := err.(kp.AlreadyLockedError); ok {
```

Comment:
The short explanation is that if a lock succeeds, I assume the key will stay assigned to the client's session and will only change for two reasons: the client explicitly unlocked it or the session expired.  Until one of those two things happens, a repeated attempt to unlock the session will eventually succeed.  If it succeeds, the method exits.  If the session expires, this will be noticed outside the process and the `done` channel will be signaled, and the method exits.